tests/tools/k_means_test.py

```diff
@@ -0,0 +1,378 @@
+    x[:, 1] += d
+    y = np.random.normal(size=(n, ndim))
+    y[:, 0] -= d
+    y[:, 1] -= d
+    z = np.random.normal(size=(n, ndim))
+    z[:, 0] += d
+    z[:, 1] -= d
+    w = np.random.normal(size=(n, ndim))
+    w[:, 0] -= d
+    w[:, 1] += d
+    x = jnp.concatenate((x, y, z, w))
+    # ignore `x` by setting its weights to 0
+    weights = jnp.ones((x.shape[0],)).at[:n].set(0.)
+
+    expected_centroids = jnp.stack([w.mean(0), z.mean(0), y.mean(0)])
+    res = k_means.k_means(x, k=k, weights=weights, key=rng)
+
+    cost = pointcloud.PointCloud(res.centroids, expected_centroids).cost_matrix
+    ixs = jnp.argmin(cost, axis=1)
+
+    np.testing.assert_array_equal(jnp.sort(ixs), jnp.arange(k))
+
+    total_shift = jnp.sum(cost[jnp.arange(k), ixs])
+    np.testing.assert_allclose(total_shift, 0., rtol=1e-3, atol=1e-3)
+
+  def test_cosine_cost_fn(self):
+    k = 4
+    geom, _, _ = make_blobs(n_samples=75)
+    geom_scaled = pointcloud.PointCloud(geom * 10., cost_fn=costs.Cosine())
+    geom = pointcloud.PointCloud(geom, cost_fn=costs.Cosine())
+
+    res_scaled = k_means.k_means(geom_scaled, k=k)
+    res = k_means.k_means(geom, k=k)
+
+    np.testing.assert_allclose(
+        res_scaled.error, res.error, rtol=1e-5, atol=1e-5
+    )
+    assert _is_same_clustering(
+        np.array(res_scaled.assignment), np.array(res.assignment), k
+    )
+
+  @pytest.mark.fast.with_args("init", ["k-means++", "random"], only_fast=0)
+  def test_k_means_jitting(
+      self, rng: jnp.ndarray, init: Literal["k-means++", "random"]
+  ):
+
+    def callback(x: jnp.ndarray) -> k_means.KMeansOutput:
+      return k_means.k_means(
+          x, k=k, init=init, store_inner_errors=True, key=rng
+      )
+
+    k = 7
+    x, _, _ = make_blobs(n_samples=150, centers=k, random_state=0)
+    res = jax.jit(callback)(x)
+    res_jit: k_means.KMeansOutput = jax.jit(callback)(x)
+
+    np.testing.assert_allclose(res.centroids, res_jit.centroids)
+    np.testing.assert_array_equal(res.assignment, res_jit.assignment)
+    np.testing.assert_allclose(res.error, res_jit.error)
+    np.testing.assert_allclose(res.inner_errors, res_jit.inner_errors)
+    assert res.iteration == res_jit.iteration
+    assert res.converged == res_jit.converged
+
+  def test_k_means_differentiability(self, rng: jnp.ndarray):
+
+    def callback(x: jnp.ndarray, w: jnp.ndarray) -> float:
+      return k_means.k_means(
+          x, k=k, weights=w, min_iterations=10, max_iterations=10, key=key1
+      ).error
+
+    k, step_size = 4, 0.1
+    x, _, _ = make_blobs(n_samples=150, centers=k, random_state=41)
+    key1, key2 = jax.random.split(rng)
+    weights = jnp.abs(jax.random.normal(key2, (x.shape[0],)))
+
+    inertia, (grad_x, grad_w) = jax.value_and_grad(callback, (0, 1))(x, weights)
+
+    inertia_x_step = callback(x - step_size * grad_x, weights)
+    inertia_w_step = callback(x, weights - step_size * weights)
+
+    assert inertia_x_step < inertia
+    assert inertia_w_step < inertia
```

Comment:
Here it seems you are testing
``iner(x - ∇iner(x)) < iner(x)`` for differentiability, but wouldn't it be a bit more convincing to have the more usual

``iner(x + v) - iner(x - v) ~= 2 <v, ∇iner(x)>``

for a small ``v``?

Reply:
Thank you, done!